src/index.js

```diff
@@ -406,69 +451,83 @@ export default class Critters {
 
     // Walk all CSS rules, marking unused rules with `.$$remove=true` for removal in the second pass.
     // This first pass is also used to collect font and keyframe usage used in the second pass.
-    walkStyleRules(ast, markOnly(rule => {
-      if (rule.type === 'rule') {
-        // Filter the selector list down to only those match
-        rule.filterSelectors(sel => {
-          // Strip pseudo-elements and pseudo-classes, since we only care that their associated elements exist.
-          // This means any selector for a pseudo-element or having a pseudo-class will be inlined if the rest of the selector matches.
-          if (sel !== ':root') {
-            sel = sel.replace(/(?:>\s*)?::?[a-z-]+\s*(\{|$)/gi, '$1').trim();
-          }
-          if (!sel) return false;
+    walkStyleRules(
+      ast,
+      markOnly((rule) => {
+        if (rule.type === 'rule') {
+          // Filter the selector list down to only those match
+          rule.filterSelectors((sel) => {
+            // Strip pseudo-elements and pseudo-classes, since we only care that their associated elements exist.
+            // This means any selector for a pseudo-element or having a pseudo-class will be inlined if the rest of the selector matches.
+            if (sel !== ':root') {
+              sel = sel.replace(/(?:>\s*)?::?[a-z-]+\s*(\{|$)/gi, '$1').trim();
+            }
+            if (!sel) return false;
 
-          try {
-            return document.querySelector(sel) != null;
-          } catch (e) {
-            failedSelectors.push(sel + ' -> ' + e.message);
+            try {
+              return document.querySelector(sel) != null;
```

Comment:
I wonder if this could be optimized in the future to reduce the amount of dom traversals. As selectors match I think one could keep them in a cache (split by space I guess) and somehow do a lookup for partial matches eg.

```
.foo .bar div // matched sometime before current selector
.foo div // current selector -> no need to query the dom as this will match 100%
```

Probably something like this (haven't thought this through so I may be ridiculously wrong :)

```js
const cache = {
  // .foo .bar div matched at some point
  div: new Set(['.foo', '.bar',])    
}

function match(sel) {
  const ancestors = sel.split(' ').reverse()
  const leaf = ancestors.shift()
  const cached = cache[leaf]
  return {
    matches: (
      cached && 
      ancestors.every(sel => cached.has(sel))
    ),
    leaf,
    ancestors
   }
}

match('.foo div').matches // true
match('div').matches // true
match('.foo .bar .baz div').matches // false -> querySelector -> if matches add `ancestors` to `cache[leaf]`
```